### DISCUSSION TITLE
Can't wake esp8266 devkit from deepsleep using RTC

### DISCUSSION THREAD

**Discussion by Alsele** [micropython/micropython#15167] [ESP8266]
I am a first time user of esp8266, I have been testing it in preparation to use it in connection with a soil moisture at home.
The board seems to work fine. REPL works fine when connected to serial (using picocom) webREPL also works. Several other modules work as expected (e.g. using sockets to write to local server, blinking LEDs, reading digital sensor, etc).

However, I am at a loss as to how to get it to wake up properly after a deep sleep. In a nutshell, I'm following the [documentation here](https://docs.micropython.org/en/latest/esp8266/quickref.html#deep-sleep-mode), However, when the defined sleep period elapses, the board flashes the small LED that is on the ESP8266 itself (not the kit) but does not continue with the boot process, i.e. I don't see the test LED going ON and OFF. 

**It only restarts after connecting the RST pin to any GND pin and then disconnecting it.** In this case it boots up and runs the boot.py scrip indicated by the LED flashing and when connected to serial it prints 5 when `machine.reset_cause()` and returns True for `if machine.reset_cause() == machine.DEEPSLEEP_RESET:` 
Any help is appreciated.

### My set up
Below are more detailed description of the set up I have, the behaviour that I am observing and the online resources that I have checked.

1- The board: [AZ-Delivery NodeMCU LUA Amica V2](https://www.az-delivery.de/en/products/nodemcu) ([pinout diagram](https://cdn.shopify.com/s/files/1/1509/1638/files/NodeMCU_LUA_Amica_V2_Pinout_Diagram.pdf?14596320546790113351))

2- Firmware: micropython ESP8266 Port, latest release [v1.22.2 (2024-02-22) .bin](https://www.micropython.org/resources/firmware/ESP8266_GENERIC-20240222-v1.22.2.bin) 
Flashed using esptool and I am using ampy to upload python scripts while the board is connected to USB port (/dev/ttyUSB0)

3- The circuit I have to test the deepsleep code is simple, connecting D0 to RST and wiring up an LED and a resistor for visual aid to check if board is rebooting properly when it is not connected to a serial port. The image at the bottom shows this circuit.

4- My custom script to go into deepsleep and wake up using RTC was added to boot.py, added below. In the beginning I had it as a separate main.py and the board behaved the same way, I added my scrip to boot.py as a precaution in case there is an issue in filesystem.

5- Not sure if relevant, but I'm using Debian Bookworm

### Resources that I checked and tutorials that I tried:
- [Deep-sleep mode in micropython documentation](https://docs.micropython.org/en/latest/esp8266/quickref.html#deep-sleep-mode)
- [The black magic of returning from Deep Sleep every time](https://www.esp8266.com/viewtopic.php?f=166&t=20500) this one mentions using a diode between D0 and RST but I don't know how to.
- [MicroPython: ESP8266 Deep Sleep and Wake Up Sources](https://randomnerdtutorials.com/micropython-esp8266-deep-sleep-wake-up-sources/)
- [ESP8266 Deep Sleep and Wake Up Sources using MicroPython](https://microcontrollerslab.com/esp8266-deep-sleep-wake-up-sources-micropython/)
- Eventually, most tutorials are the same, connecting D0 to RST directly, define a deep_sleep function with RTC timer.

I tried adding a resistor (10k ohm) between D0 and RST but that does not make the board wake up properly. However, it changes the behaviour so that a small LED on the board, close from the RST pin turns ON constantly. It also stays ON after disconnecting the RST pin from the 10k resistor. Rebooting the board in this case, requires the RST pin to be connected and disconnected twice to a GND pin.

I'm also starting to suspect that this esp8266 kit does not have the capability to wake up from deep sleep using RTC, because the pinout diagram linked above does not show a "WAKE" tag next to D0, although the schematic (what's the difference?) does show that [schematic](https://cdn.shopify.com/s/files/1/1509/1638/files/NodeMCU_LUA_Amica_V2_Pinout_Diagram.pdf?14596320546790113351)

I would really appreciate any help to figure this out, it's been 3 days...

### boot.py script

```py
# This file is executed on every boot (including wake-boot from deepsleep)
#import esp
#esp.osdebug(None)
import os, machine
#os.dupterm(None, 1) # disable REPL on UART(0)
import gc
#import webrepl
#webrepl.start()
gc.collect()

from machine import Pin
import network
from time import sleep

Pin(5, Pin.OUT).on()
sleep(5)
Pin(5, Pin.OUT).off()
sleep(1)

print(machine.reset_cause())
if machine.reset_cause() == machine.DEEPSLEEP_RESET:
    print('woke from a deep sleep')


def deep_sleep(msec):
    ap= network.WLAN(network.AP_IF)
    if ap.active():
        ap.active(False)
        sleep(3)

    #configure RTC.ALARM
    rtc= machine.RTC()
    rtc.irq(trigger=rtc.ALARM0, wake=machine.DEEPSLEEP)

    rtc.alarm(rtc.ALARM0, msec)

    machine.deepsleep()
    sleep(1)

i=0
led= Pin(5, Pin.OUT)
while i < 3:
    led.on()
    sleep(2)
    led.off()
    sleep(2)
    i= i+1

print("going to sleep in 10 seconds for 5 seconds")
sleep(10)

deep_sleep(5000)
```

![AZDelivery-ESP8266MOD-12-F](https://github.com/micropython/micropython/assets/142629233/e9cf74f2-3b2f-483d-8e23-1a8fb63d6606)

**Comment by davefes** [comment]
Try:
1) remove all code from boot.py
2) place this in `main.py` to give you  chance at breaking-in before code execution
```py
import time
import machine


#  during development
if machine.reset_cause() == machine.SOFT_RESET:
    print('you got a SOFT_RESET, so do a machine.reset()')
    time.sleep(1)
    machine.reset()

print('wait 2 seconds for a CTRL-C')
time.sleep(2)

import my_program
```
3) put all your code in `my_program`

BTW, the sleep(1) after calling deepsleep() is never run. 

Remove the 2 second delay after getting things to run properly, if it is not wanted.

**Reply by Alsele** [reply]
Thanks a lot, I tested your proposed solution but, sadly, the esp continues to behave in the same manner.

I flashed a similar board with NodeMCU and used `rtctime.dsleep` and it worked as expected. So I'm ruling out the board being defected or having other hardware issues.

I am now suspecting `machine.RTC`, either I'm not using it correctly, despite following the example in the documentation, or that it is not documented sufficiently, or ultimately, that it is not implemented properly? (this is a stretch because I'm not qualified to make such a judgment)

I found this [stackoverflow](https://stackoverflow.com/questions/49757733/micropython-on-esp8266-rtc-does-not-have-an-init-method) stating errors exist in the documentation from six years ago and strangely it seems to still be valid today. See below from the latest micropython firmware. [`RTC.init()`](https://docs.micropython.org/en/v1.22.0/library/machine.RTC.html#machine.RTC.init) is missing.
```py
>>> dir(machine.RTC)
['__class__', '__name__', 'ALARM0', '__bases__', '__dict__', 'alarm', 'alarm_left', 'datetime', 'irq', 'memory']
```

I also noticed, contrary to [documentation](https://docs.micropython.org/en/v1.22.0/library/machine.RTC.html#machine.RTC.alarm), that `RTC.alarm(id, time, *, repeat=False)` does not accept the keyword `repeat`
```py
>>> rtc= machine.RTC()
>>> rtc.alarm(5000, repeat=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: function doesn't take keyword arguments
```
nor as argument.
```py
>>> rtc.alarm(5000, True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: invalid alarm
```

For my use case, being very simple, I might be able to hack few lines of code in Lua for NodeMCU, despite not having any previous knowledge, and get on with my project. But I would really hope that this can be done in micropython.
